src/learning/toolkit/net_layer.rs

```diff
@@ -0,0 +1,151 @@
+//!Neural Network Layers
+
+use linalg::{Matrix, MatrixSlice};
+use linalg::BaseSlice;
+
+use learning::toolkit::activ_fn::ActivationFunc;
+
+use rand::thread_rng;
+use rand::distributions::Sample;
+use rand::distributions::normal::Normal;
+
+use std::fmt::Debug;
+
+/// Trait for neural net layers
+pub trait NetLayer : Debug {
+	/// The result of propogating data forward through this layer
+	fn forward(&self, input: &Matrix<f64>, params: MatrixSlice<f64>) -> Matrix<f64>;
+
+	/// The gradient of the output of this layer with respect to its input
+	fn back_input(&self, out_grad: &Matrix<f64>, input: &Matrix<f64>, params: MatrixSlice<f64>) -> Matrix<f64>;
+	
+	/// The gradient of the output of this layer with respect to its parameters
+	fn back_params(&self, out_grad: &Matrix<f64>, input: &Matrix<f64>, params: MatrixSlice<f64>) -> Matrix<f64>;
+
+	/// The default value of the parameters of this layer before training
+	fn default_params(&self) -> Vec<f64>;
+
+	/// The number of parameters used by this layer
+	fn num_params(&self) -> usize;
+
+	/// The shape of the parameters used by this layer
+	fn param_shape(&self) -> (usize, usize);
+}
+
+/// Linear network layer
+///
+/// Represents a fully connected layer with optional bias term
+///
+/// The parameters are a matrix of weights of size I x O
+/// where O is the dimensionality of the output and I the dimensionality of the input
+#[derive(Debug, Clone, Copy)]
+pub struct Linear {
+	/// The number of dimensions of the input
+	input_size: usize,
+	/// The number of dimensions of the output
+	output_size: usize,
+	/// Whether or not to include a bias term
+	has_bias: bool,
+}
+
+impl Linear {
+	/// Construct a new Linear layer
+	pub fn with_bias(input_size: usize, output_size: usize) -> Linear {
```

Comment:
It's not your fault at all! I just can't help being super picky with user-facing things (like docs and public function names). Feel free to let me know if you disagree (I'll be more than happy to hear you out). Also you can hang onto these minor changes until later if you prefer.